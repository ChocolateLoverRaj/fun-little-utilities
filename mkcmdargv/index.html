<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>mkcmdargv — Reference</title>
	<meta name="description" content="Documentation for mkcmdargv — Reference">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="assets/css/main.css">
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="assets/js/search.json" data-base=".">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="index.html" class="title">mkcmdargv — Reference</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="globals.html">Globals</a>
				</li>
			</ul>
			<h1>mkcmdargv — Reference</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<div class="tsd-panel tsd-typography">
				<a href="#create-command-arguments" id="create-command-arguments" style="color: inherit; text-decoration: none;">
					<h1>Create Command Arguments</h1>
				</a>
				<p>Convert environment variables to command-line arguments</p>
				<a href="#executable-usage" id="executable-usage" style="color: inherit; text-decoration: none;">
					<h2>Executable Usage</h2>
				</a>
				<pre><code class="language-sh"><span class="hljs-built_in">export</span> MKCMDARGV_ARGS=<span class="hljs-string">'[abc def]'</span>
<span class="hljs-built_in">export</span> MKCMDARGV_OPTIONS=<span class="hljs-string">'{ a: 0, foo: 123 }'</span>
<span class="hljs-built_in">export</span> MKCMDARGV_SINGLE_FLAGS=<span class="hljs-string">'sfx'</span>
<span class="hljs-built_in">export</span> MKCMDARGV_DOUBLE_FLAGS=<span class="hljs-string">'quiet splash'</span>
<span class="hljs-built_in">echo</span> $(mkcmdargv)</code></pre>
				<a href="#api-usage" id="api-usage" style="color: inherit; text-decoration: none;">
					<h2>API Usage</h2>
				</a>
				<a href="#iteratecommandarguments" id="iteratecommandarguments" style="color: inherit; text-decoration: none;">
					<h3><code>iterateCommandArguments</code></h3>
				</a>
				<p>Creates an iterator of command arguments</p>
				<pre><code class="language-javascript"><span class="hljs-keyword">const</span> { iterateCommandArguments } = <span class="hljs-built_in">require</span>(<span class="hljs-string">'mkcmdargv'</span>)

<span class="hljs-built_in">console</span>.log(...iterateCommandArguments({
  <span class="hljs-attr">args</span>: [<span class="hljs-string">'abc'</span>, <span class="hljs-string">'def'</span>],
  <span class="hljs-attr">options</span>: { <span class="hljs-attr">a</span>: <span class="hljs-number">0</span>, <span class="hljs-attr">foo</span>: <span class="hljs-number">123</span> },
  <span class="hljs-attr">flags</span>: [<span class="hljs-string">'s'</span>, <span class="hljs-string">'f'</span>, <span class="hljs-string">'x'</span>, <span class="hljs-string">'quiet'</span>, <span class="hljs-string">'splash'</span>]
}))</code></pre>
				<a href="#createcommandarguments" id="createcommandarguments" style="color: inherit; text-decoration: none;">
					<h3><code>createCommandArguments</code></h3>
				</a>
				<p>Creates an array of command arguments</p>
				<pre><code class="language-javascript"><span class="hljs-keyword">const</span> { createCommandArguments } = <span class="hljs-built_in">require</span>(<span class="hljs-string">'mkcmdargv'</span>)

createCommandArguments({
  <span class="hljs-attr">args</span>: [<span class="hljs-string">'abc'</span>, <span class="hljs-string">'def'</span>],
  <span class="hljs-attr">options</span>: { <span class="hljs-attr">a</span>: <span class="hljs-number">0</span>, <span class="hljs-attr">foo</span>: <span class="hljs-number">123</span> },
  <span class="hljs-attr">flags</span>: [<span class="hljs-string">'s'</span>, <span class="hljs-string">'f'</span>, <span class="hljs-string">'x'</span>, <span class="hljs-string">'quiet'</span>, <span class="hljs-string">'splash'</span>]
}).forEach(<span class="hljs-function"><span class="hljs-params">arg</span> =&gt;</span> <span class="hljs-built_in">console</span>.log(arg))</code></pre>
				<a href="#license" id="license" style="color: inherit; text-decoration: none;">
					<h2>License</h2>
				</a>
				<p><a href="https://git.io/fxKXN">MIT</a> © <a href="https://github.com/KSXGitHub">Hoàng Văn Khải</a></p>
			</div>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class="globals  ">
						<a href="globals.html"><em>Globals</em></a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
					<li class=" tsd-kind-variable">
						<a href="globals.html#bin" class="tsd-kind-icon">bin</a>
					</li>
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="assets/js/main.js"></script>
</body>
</html>