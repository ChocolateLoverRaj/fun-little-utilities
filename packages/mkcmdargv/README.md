# Create Command Arguments

Convert environment variables to command-line arguments

## Executable Usage

```sh
export MKCMDARGV_ARGS='[abc def]'
export MKCMDARGV_OPTIONS='{ a: 0, foo: 123 }'
export MKCMDARGV_SINGLE_FLAGS='sfx'
export MKCMDARGV_DOUBLE_FLAGS='quiet splash'
echo $(mkcmdargv)
```

## API Usage

### `iterateCommandArguments`

Creates an iterator of command arguments

```javascript
const { iterateCommandArguments } = require('mkcmdargv')

console.log(...iterateCommandArguments({
  args: ['abc', 'def'],
  options: { a: 0, foo: 123 },
  flags: ['s', 'f', 'x', 'quiet', 'splash']
}))
```

### `createCommandArguments`

Creates an array of command arguments

```javascript
const { createCommandArguments } = require('mkcmdargv')

createCommandArguments({
  args: ['abc', 'def'],
  options: { a: 0, foo: 123 },
  flags: ['s', 'f', 'x', 'quiet', 'splash']
}).forEach(arg => console.log(arg))
```

## License

[MIT](https://git.io/fxKXN) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
