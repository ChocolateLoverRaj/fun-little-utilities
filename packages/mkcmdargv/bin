#! /usr/bin/env node
require('mkcmdargv-cli/main')
