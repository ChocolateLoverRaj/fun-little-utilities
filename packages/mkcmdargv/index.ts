export * from './lib'
export const bin = require.resolve('./bin/executable')
