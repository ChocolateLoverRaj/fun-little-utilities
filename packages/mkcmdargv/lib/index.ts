import { partition } from 'ramda'
import { splitIterable } from 'split-iterable'

type Value = string | number | boolean

type Options = {
  readonly [name in string]: Value
}

export interface Param {
  readonly args?: readonly string[]
  readonly options?: Options
  readonly flags?: readonly string[]
}

function parseSingleDash ([key, value]: [string, Value]) {
  const optName = '-' + key
  if (value === true) return [optName]
  if (value === false) return []
  return [optName, String(value)]
}

function parseDoubleDash ([key, value]: [string, Value]) {
  const optName = '--' + key
  if (value === true) return [optName]
  if (value === false) return []
  return [`${optName}=${value}`]
}

export function * iterateCommandArguments (param: Param) {
  const { args = [], options = {}, flags = [] } = param
  const [beforeDoubleDashes, ...afterDoubleDashes] = splitIterable(args, x => x === '--')
  const [beforeDashes, afterDashes] = partition(x => x[0] !== '-', beforeDoubleDashes)
  const optionsPairs = Object.entries(options)
  const [singleDashOptions, doubleDashOptions] = partition(([key]) => key.length === 1, optionsPairs)
  const [singleDashFlags, doubleDashFlags] = partition(x => x.length === 1, flags)

  yield * beforeDashes

  if (singleDashFlags.length) yield '-' + singleDashFlags.join('')

  for (const entry of singleDashOptions) {
    yield * parseSingleDash(entry)
  }

  yield * doubleDashFlags.map(x => '--' + x)

  for (const entry of doubleDashOptions) {
    yield * parseDoubleDash(entry)
  }

  if (!afterDashes.length && !afterDoubleDashes.length) return

  yield '--'
  yield * afterDashes

  for (const chunk of afterDoubleDashes) {
    yield '--'
    yield * chunk
  }
}

export function createCommandArguments (param: Param) {
  return Array.from(iterateCommandArguments(param))
}

export async function help () {
  const pathPromise = import('path')
  const fsxPromise = import('fs-extra')
  const path = await pathPromise
  const helpFilePath = path.resolve(__dirname, '../data/help.txt')
  const { readFile } = await fsxPromise
  return readFile(helpFilePath, { encoding: 'utf8' })
}
