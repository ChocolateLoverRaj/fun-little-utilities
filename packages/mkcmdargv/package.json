{
  "name": "mkcmdargv",
  "version": "0.1.0",
  "description": "Convert environment variables to command-line arguments",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/ksxnodeapps/fun-little-utilities#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "bin": {
    "mkcmdargv": "./bin/executable",
    "create-command-arguments": "./bin/executable"
  },
  "dependencies": {
    "ramda": "^0.26.1",
    "js-yaml": "^3.13.1",
    "fs-extra": "^8.1.0",
    "split-iterable": "^0.1.0",
    "tslib": "^1.10.0",
    "@types/ramda": "^0.26.21",
    "@types/js-yaml": "^3.12.1",
    "@types/fs-extra": "^8.0.0",
    "@types/node": "^12.7.8"
  }
}
