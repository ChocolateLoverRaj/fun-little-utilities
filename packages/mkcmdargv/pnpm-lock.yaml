dependencies:
  '@types/fs-extra': 8.0.0
  '@types/js-yaml': 3.12.1
  '@types/node': 12.7.8
  '@types/ramda': 0.26.25
  fs-extra: 8.1.0
  js-yaml: 3.13.1
  ramda: 0.26.1
  split-iterable: 'link:../split-iterable'
  tslib: 1.10.0
lockfileVersion: 5.1
packages:
  /@types/fs-extra/8.0.0:
    dependencies:
      '@types/node': 12.7.2
    dev: false
    resolution:
      integrity: sha512-bCtL5v9zdbQW86yexOlXWTEGvLNqWxMFyi7gQA7Gcthbezr2cPSOb8SkESVKA937QD5cIwOFLDFt0MQoXOEr9Q==
  /@types/js-yaml/3.12.1:
    dev: false
    resolution:
      integrity: sha512-SGGAhXLHDx+PK4YLNcNGa6goPf9XRWQNAUUbffkwVGGXIxmDKWyGGL4inzq2sPmExu431Ekb9aEMn9BkPqEYFA==
  /@types/node/12.7.2:
    dev: false
    resolution:
      integrity: sha512-dyYO+f6ihZEtNPDcWNR1fkoTDf3zAK3lAABDze3mz6POyIercH0lEUawUFXlG8xaQZmm1yEBON/4TsYv/laDYg==
  /@types/node/12.7.8:
    dev: false
    resolution:
      integrity: sha512-FMdVn84tJJdV+xe+53sYiZS4R5yn1mAIxfj+DVoNiQjTYz1+OYmjwEZr1ev9nU0axXwda0QDbYl06QHanRVH3A==
  /@types/ramda/0.26.25:
    dependencies:
      ts-toolbelt: 4.5.2
    dev: false
    resolution:
      integrity: sha512-fx54Gx7Xrw+wMU7sEbUa5tjgL5Ec/vWzhKCqJHR0JVJUqwGzUtIYk8nqwyLbmX11wEx2xVdblYlNL9BzznDhvw==
  /argparse/1.0.10:
    dependencies:
      sprintf-js: 1.0.3
    dev: false
    resolution:
      integrity: sha512-o5Roy6tNG4SL/FOkCAN6RzjiakZS25RLYFrcMttJqbdd8BWrnA+fGz57iN5Pb06pvBGvl5gQ0B48dJlslXvoTg==
  /esprima/4.0.1:
    dev: false
    engines:
      node: '>=4'
    hasBin: true
    resolution:
      integrity: sha512-eGuFFw7Upda+g4p+QHvnW0RyTX/SVeJBDM/gCtMARO0cLuT2HcEKnTPvhjV6aGeqrCB/sbNop0Kszm0jsaWU4A==
  /fs-extra/8.1.0:
    dependencies:
      graceful-fs: 4.2.2
      jsonfile: 4.0.0
      universalify: 0.1.2
    dev: false
    engines:
      node: '>=6 <7 || >=8'
    resolution:
      integrity: sha512-yhlQgA6mnOJUKOsRUFsgJdQCvkKhcz8tlZG5HBQfReYZy46OwLcY+Zia0mtdHsOo9y/hP+CxMN0TU9QxoOtG4g==
  /graceful-fs/4.2.2:
    dev: false
    resolution:
      integrity: sha512-IItsdsea19BoLC7ELy13q1iJFNmd7ofZH5+X/pJr90/nRoPEX0DJo1dHDbgtYWOhJhcCgMDTOw84RZ72q6lB+Q==
  /js-yaml/3.13.1:
    dependencies:
      argparse: 1.0.10
      esprima: 4.0.1
    dev: false
    hasBin: true
    resolution:
      integrity: sha512-YfbcO7jXDdyj0DGxYVSlSeQNHbD7XPWvrVWeVUujrQEoZzWJIRrCPoyk6kL6IAjAG2IolMK4T0hNUe0HOUs5Jw==
  /jsonfile/4.0.0:
    dev: false
    optionalDependencies:
      graceful-fs: 4.2.2
    resolution:
      integrity: sha1-h3Gq4HmbZAdrdmQPygWPnBDjPss=
  /ramda/0.26.1:
    dev: false
    resolution:
      integrity: sha512-hLWjpy7EnsDBb0p+Z3B7rPi3GDeRG5ZtiI33kJhTt+ORCd38AbAIjB/9zRIUoeTbE/AVX5ZkU7m6bznsvrf8eQ==
  /sprintf-js/1.0.3:
    dev: false
    resolution:
      integrity: sha1-BOaSb2YolTVPPdAVIDYzuFcpfiw=
  /ts-toolbelt/4.5.2:
    dev: false
    resolution:
      integrity: sha512-g7ZDTK17HhZi5qhFvHLaW/ZzaOlEB+O7GnDaW6ftCEaxhbenVEzPfmeGo2YhMgRjhlYO/dtOmUbv0yykQU5YMw==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
  /universalify/0.1.2:
    dev: false
    engines:
      node: '>= 4.0.0'
    resolution:
      integrity: sha512-rBJeI5CXAlmy1pV+617WB9J63U6XcazHHF2f2dbJix4XzpUF0RS3Zbj0FGIOCAva5P/d/GBOYaACQ1w+0azUkg==
specifiers:
  '@types/fs-extra': ^8.0.0
  '@types/js-yaml': ^3.12.1
  '@types/node': ^12.7.8
  '@types/ramda': ^0.26.25
  fs-extra: ^8.1.0
  js-yaml: ^3.13.1
  ramda: ^0.26.1
  split-iterable: ^0.1.1
  tslib: ^1.10.0
