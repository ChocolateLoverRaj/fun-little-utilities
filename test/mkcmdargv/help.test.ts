import { help } from 'mkcmdargv'

it('matches snapshot', async () => {
  expect('\n' + await help()).toMatchSnapshot()
})
