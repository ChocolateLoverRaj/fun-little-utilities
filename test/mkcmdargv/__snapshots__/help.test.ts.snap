// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
"
Usage:
  mkcmdargv [--help]

Options:
  --help: Show this message

Environment Variables:
  MKCMDARGV_ARGS: YAML code of args array (e.g. '[123, 456, abc, def]')
  MKCMDARGV_OPTIONS: YAML code of options object (e.g. '{ a: 0, foo: 123 }')
  MKCMDARGV_SINGLE_FLAGS: String of single-character flags (e.g. 'sfx')
  MKCMDARGV_DOUBLE_FLAGS: Whitespace separated list of multi-character flags: (e.g. 'quiet splash')
"
`;
